pallets/asset/src/checkpoint.rs

```diff
@@ -0,0 +1,551 @@
+// the Free Software Foundation, version 3.
+
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+//! # Checkpoint Module
+//!
+//! The Checkpoint module provides extrinsics and storage to take snapshots,
+//! henceforth called *checkpoints*, of the supply of assets,
+//! and how they were distributed at the time of checkpoint.
+//!
+//! Using the module, users can also schedule checkpointing in the future,
+//! either at fixed points in time (e.g., "next friday"),
+//! or at regular intervals (e.g., "every month").
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Functions
+//!
+//! - `create_checkpoint` creates a checkpoint.
+//! - `create_schedule` creates a checkpoint schedule.
+//! - `remove_schedule` removes a checkpoint schedule.
+//!
+//! ### Public Functions
+//!
+//! - `balance_at(ticker, did, cp)` returns the balance of `did` for `ticker` at least `>= cp`.
+//! - `advance_update_balances(ticker, updates)` advances schedules for `ticker`
+//!    and applies new balances in `updates` for the last checkpoint.
+//! - Other misc storage items as defined in `decl_storage!`.
+
+use codec::{Decode, Encode};
+use core::mem;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage,
+    dispatch::{DispatchError, DispatchResult},
+    ensure,
+    traits::{Get, UnixTime},
+};
+use frame_system::ensure_root;
+use polymesh_common_utilities::{
+    protocol_fee::{ChargeProtocolFee, ProtocolOp},
+    CommonTrait, GC_DID,
+};
+use polymesh_primitives::{
+    calendar::{CheckpointId, CheckpointSchedule},
+    IdentityId, Moment, Ticker,
+};
+use sp_runtime::traits::SaturatedConversion;
+#[cfg(feature = "std")]
+use sp_runtime::{Deserialize, Serialize};
+use sp_std::prelude::*;
+
+use crate as pallet_asset;
+use pallet_asset::Trait;
+
+type Asset<T> = pallet_asset::Module<T>;
+
+/// ID of a `StoredSchedule`.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Copy, Clone, Debug, PartialEq, Eq, Default)]
+pub struct ScheduleId(pub u64);
+
+/// One or more scheduled checkpoints in the future.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Copy, Clone, Debug, PartialEq, Eq)]
+pub struct StoredSchedule {
+    /// A series of checkpoints in the future defined by the schedule.
+    pub schedule: CheckpointSchedule,
+    /// The ID of the schedule itself.
+    /// Not to be confused for a checkpoint's ID.
+    pub id: ScheduleId,
+    /// When the next checkpoint is due to be created.
+    /// Used as a cache for more efficient sorting.
+    pub at: Moment,
+}
+
+decl_storage! {
```

Comment:
The checkpoint system still needs to talk to the asset module, so it becomes simpler to have this module here rather than use a proxy trait. Also, time constraints... was quicker to put the checkpoint module here for now. We can extract it later without storage migration, as it has separate storage, events, errors, extrinsics, etc.

The UI should not care about the trait system at all, since it doesn't exist at runtime (aside from trait objects, which only manifest as vtables...).

Reply:
Yeah UI doesn't care about the trait system, I am more concerned about reading the storage variables from the UI. Have you try reading storage from the chain state tab of the UI for the checkpoint module?

Reply:
FYI - I ran a --dev node and was able to set checkpoints, and read checkpoint data through the polkadot.js UI without issue.

Reply:
![Screenshot 2020-10-28 at 09 44 31](https://user-images.githubusercontent.com/2633913/97419673-3ed97600-1902-11eb-8e8d-7802a84a1ee1.png)

Reply:
Perfect; thanks @adamdossa. Let's call this resolved?